File: Validacao de Datas/dataCalendario.go
package main

import (
    "bufio"
    "fmt"
    "os"
    "strconv"
    "strings"
)

type Data struct {
    dia int
    mes int
    ano int
    ok  bool
}

func StrToInt(cadeia string) int {
    valor, _ := strconv.Atoi(cadeia)
    return valor
}

func IntToStr(valor int) string {
    return strconv.Itoa(valor)
}

func Formatar(valor, digitos int) string {
    valorCadeia := IntToStr(valor)
    for len(valorCadeia) < digitos {
        valorCadeia = "0" + valorCadeia
    }
    return valorCadeia
}

func anoBissexto(data Data) bool {
    if data.ano%400 == 0 {
        return true
    } else if data.ano%4 == 0 && data.ano%100 != 0 {
        return true
    } else {
        return false
    }
}

func ultimoDiaMes(data Data) int {
    if data.mes == 2 {
        if anoBissexto(data) {
            return 29
        } else {
            return 28
        }
    }
    if data.mes <= 7 {
        if data.mes%2 != 0 {
            return 31
        } else {
            return 30
        }
    }
    if data.mes%2 != 0 {
        return 30
    } else {
        return 31
    }
}

func validaData(data *Data) {
    data.ok = (data.ano >=    1) &&
              (data.ano <= 9999) &&
              (data.mes >=    1) &&
              (data.mes <=   12) &&
              (data.dia >=    1) &&
              (data.dia <= ultimoDiaMes(*data)) &&
            !((data.ano == 1582) &&
              (data.mes ==   10) &&
              (data.dia >=    5) &&
              (data.dia <=   14))
}

func alteraData(data *Data, dia, mes, ano int) {
    data.dia = dia
    data.mes = mes
    data.ano = ano
    validaData(data)
}

func mostraData(data Data) {
    if data.ok {
        fmt.Printf("%s/%s/%s\n", Formatar(data.dia, 2), Formatar(data.mes, 2), Formatar(data.ano, 4))
    } else {
        fmt.Println("**/**/****")
    }
}

func main() {
    fmt.Println("TESTE DE VALIDACAO DE DATA (CALENDARIO)")
    fmt.Println("=======================================")
    fmt.Println("")

    reader := bufio.NewReader(os.Stdin)
    fmt.Print("Entre data no formato DD/MM/AAAA: ")
    DTS, _ := reader.ReadString('\n')
    DTS = strings.TrimSpace(DTS)

    DIA := StrToInt(DTS[0:2])
    MES := StrToInt(DTS[3:5])
    ANO := StrToInt(DTS[6:10])

    data := Data{}
    alteraData(&data, DIA, MES, ANO)

    fmt.Print("Data informada: ")
    mostraData(data)

    fmt.Println("")
    fmt.Print("Tecle <Enter> para encerrar... ")
    reader.ReadString('\n')
}

File: Validacao de Datas/dataCalendario.pas
program dataCalendario;

var
  DTS: string;
  DIA, MES, ANO: Integer;
  data: record
    dia: Integer;
    mes: Integer;
    ano: Integer;
    ok:  Boolean;
  end;
  
function StrToInt(const cadeia: string): Integer;
var
  I, valor, sinal: Integer;
begin
  valor := 0;
  sinal := 1;
  if cadeia = '' then
  begin
    StrToInt := 0;
    Exit;
  end;
  if cadeia[1] = '-' then
  begin
    sinal := -1;
    I := 2;
  end
  else
    I := 1;
  while I <= Length(cadeia) do
  begin
    if not (cadeia[I] in ['0'..'9']) then
    begin
      StrToInt := 0;
      Exit;
    end;
    valor := valor * 10 + Ord(cadeia[I]) - Ord('0');
    Inc(I);
  end;
  StrToInt := valor * sinal;
end;

function IntToStr(valor: Integer): string;
begin
  Str(valor, IntToStr);
end;

function Formatar(valor: Integer; digitos: Integer): string;
var
  valorCadeia: string;
begin
  valorCadeia := IntToStr(valor);
  while Length(valorCadeia) < digitos do
    valorCadeia := '0' + valorCadeia;
  Formatar := valorCadeia;
end;

function anoBissexto: Boolean;
begin
  if data.ano mod 400 = 0 then
    anoBissexto := true
  else if (data.ano mod 4 = 0) and (data.ano mod 100 <> 0) then
    anoBissexto := true
  else
    anoBissexto := false;
end;

function ultimoDiaMes: Integer;
begin
  if data.mes = 2 then
  begin
    if anoBissexto then
      ultimoDiaMes := 29
    else
      ultimoDiaMes := 28;
  end
  else if data.mes <= 7 then
  begin
    if data.mes mod 2 <> 0 then
      ultimoDiaMes := 31
    else
      ultimoDiaMes := 30;
  end
  else
  begin
    if data.mes mod 2 <> 0 then
      ultimoDiaMes := 30
    else
      ultimoDiaMes := 31;
  end;
end;

procedure validaData;
begin
  data.ok := (data.ano >=    1) and 
             (data.ano <= 9999) and
             (data.mes >=    1) and 
             (data.mes <=   12) and
             (data.dia >=    1) and 
             (data.dia <= ultimoDiaMes) and
             not ((data.ano = 1582) and 
                  (data.mes =   10) and 
                  (data.dia >=   5) and 
                  (data.dia <=  14));
end;

procedure alteraData(dia, mes, ano: Integer);
begin
  data.dia := dia;
  data.mes := mes;
  data.ano := ano;
  validaData;
end;

procedure mostraData;
begin
  if data.ok then
    writeln(Formatar(data.dia, 2) + '/' + Formatar(data.mes, 2) + '/' + Formatar(data.ano, 4))
  else
    writeln('**/**/****');
end;

begin
  data.dia := 0;
  data.mes := 0;
  data.ano := 0;
  data.ok := false;
  writeln('TESTE DE VALIDACAO DE DATA (CALENDARIO)');
  writeln('=======================================');
  writeln;
  write('Entre data no formato DD/MM/AAAA: ');
  readln(DTS);

  DIA := StrToInt(Copy(DTS, 1, 2));
  MES := StrToInt(Copy(DTS, 4, 2));
  ANO := StrToInt(Copy(DTS, 7, 4));

  alteraData(DIA, MES, ANO);

  write('Data informada: ');
  mostraData();

  writeln;
  write('Tecle <Enter> para encerrar... ');
  readln;
end.
